source/winAPI/sessionTracking.py

```diff
@@ -106,278 +99,153 @@ class WindowsTrackedSession(enum.IntEnum):
 	SESSION_TERMINATE = 11
 
 
-_toggleWindowsSessionStatePair: Dict[WindowsTrackedSession, WindowsTrackedSession] = {
-	WindowsTrackedSession.CONSOLE_CONNECT: WindowsTrackedSession.CONSOLE_DISCONNECT,
-	WindowsTrackedSession.CONSOLE_DISCONNECT: WindowsTrackedSession.CONSOLE_CONNECT,
-	WindowsTrackedSession.REMOTE_CONNECT: WindowsTrackedSession.REMOTE_DISCONNECT,
-	WindowsTrackedSession.REMOTE_DISCONNECT: WindowsTrackedSession.REMOTE_CONNECT,
-	WindowsTrackedSession.SESSION_LOGON: WindowsTrackedSession.SESSION_LOGOFF,
-	WindowsTrackedSession.SESSION_LOGOFF: WindowsTrackedSession.SESSION_LOGON,
-	WindowsTrackedSession.SESSION_LOCK: WindowsTrackedSession.SESSION_UNLOCK,
-	WindowsTrackedSession.SESSION_UNLOCK: WindowsTrackedSession.SESSION_LOCK,
-	WindowsTrackedSession.SESSION_CREATE: WindowsTrackedSession.SESSION_TERMINATE,
-	WindowsTrackedSession.SESSION_TERMINATE: WindowsTrackedSession.SESSION_CREATE,
-}
-"""
-Pair of WindowsTrackedSession, where each key has a value of the opposite state.
-e.g. SESSION_LOCK/SESSION_UNLOCK.
-"""
-
-
-def _hasLockStateBeenTracked() -> bool:
+class _WindowsLockedState(AutoPropertyObject):
 	"""
-	Checks if NVDA is aware of a session lock state change since NVDA started.
+	Class to encapsulate caching the Windows lock state.
 	"""
-	return bool(_currentSessionStates.intersection({
-		WindowsTrackedSession.SESSION_LOCK,
-		WindowsTrackedSession.SESSION_UNLOCK
-	}))
+	# Refer to AutoPropertyObject for notes on caching
+	_cache_isWindowsLocked = True
+	
+	# Typing information for auto-property _get_isWindowsLocked
+	isWindowsLocked: bool
 
+	def _get_isWindowsLocked(self) -> bool:
+		from winAPI.sessionTracking import _isWindowsLocked_checkViaSessionQuery
+		return _isWindowsLocked_checkViaSessionQuery()
+
+
+def initialize():
+	global _lockStateTracker
+	_lockStateTracker = _WindowsLockedState()
 
-def _recordLockStateTrackingFailure(error: Optional[Exception] = None):
-	log.error(
-		"Unknown lock state, unexpected, potential security issue, please report.",
-		exc_info=error
-	)  # Report error repeatedly, attention is required.
-	##
-	# For security it would be best to treat unknown as locked.
-	# However, this would make NVDA unusable.
-	# Instead, the user should be warned, and allowed to mitigate the impact themselves.
-	# Reporting is achieved via _sessionQueryLockStateHasBeenUnknown exposed with
-	# L{hasLockStateBeenUnknown}.
-	global _sessionQueryLockStateHasBeenUnknown
-	_sessionQueryLockStateHasBeenUnknown = True
+
+def pumpAll():
+	"""Used to track the session lock state every core cycle, and detect changes."""
+	global _wasLockedPreviousPumpAll
+	from utils.security import postSessionLockStateChanged, _searchForLockAppModule
+	windowsIsNowLocked = _isWindowsLocked()
+	if windowsIsNowLocked != _wasLockedPreviousPumpAll:
+		_wasLockedPreviousPumpAll = windowsIsNowLocked
+		postSessionLockStateChanged.notify(isNowLocked=windowsIsNowLocked)
+	_searchForLockAppModule()

```

Comment:
Would it not be better for the appModule search to happen before firing the notification? So that any callbacks can guarantee that isBelowLockscreen checks will work?

Reply:
make sense